### PR TITLE
[BEAM-2306] Fail build when @Deprecated is used without @deprecated javadoc

### PR DESCRIPTION
**Pull request by alex-filatov**
Add checkstyle check to fail the build when `@Deprecated` is used without `@deprecated` javadoc (or vice versa).

The check is disabled for existing violations where reason for deprecation and/or alternative is not clear.

### CODE REVIEW

**Comment by alex-filatov** [comment]
R: @kennknowles

File: sdks/java/core/src/main/java/org/apache/beam/sdk/testing/StreamingIT.java [hunk 1]
```diff
@@ -30,6 +30,8 @@
  *     }
  * </code></pre>
  */
+// CHECKSTYLE.OFF: MissingDeprecated
```
Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/testing/StreamingIT.java]:
```java
@deprecated tests which use unbounded PCollections should be in the category {@link UsesUnboundedPCollections}. Beyond that, it is up to the runner and test configuration to decide whether to run in streaming mode.
```

File: sdks/java/core/src/main/java/org/apache/beam/sdk/util/IdentityWindowFn.java [hunk 2]
```diff
@@ -111,7 +111,9 @@ public void verifyCompatibility(WindowFn<?, ?> other) throws IncompatibleWindowE
             getClass().getCanonicalName()));
   }
 
+  // CHECKSTYLE.OFF: MissingDeprecated
   @Deprecated
+  // CHECKSTYLE.ON: MissingDeprecated
```
Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/util/IdentityWindowFn.java]: I actually don't think this should be deprecated.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/values/PCollectionViews.java [hunk 3]
```diff
@@ -186,7 +186,9 @@ private SingletonViewFn(boolean hasDefault, T defaultValue, Coder<T> valueCoder)
     /**
      * Returns if a default value was specified.
      */
+    // CHECKSTYLE.OFF: MissingDeprecated
```
Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/values/PCollectionViews.java]:
```java
@deprecated these methods are all expected to change arbitrarily
```

But really, `@Internal` is probably enough.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/coders/CoderRegistry.java [hunk 4]
```diff
@@ -255,7 +257,9 @@ public void registerCoderForType(TypeDescriptor<?> type, Coder<?> coder) {
    *
    * @throws CannotProvideCoderException if a {@link Coder} cannot be provided
    */
+  // CHECKSTYLE.OFF: MissingDeprecated
   @Deprecated
+  // CHECKSTYLE.ON: MissingDeprecated
```
Comment (lukecwik) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/coders/CoderRegistry.java]: There is no intended alternative, this is a visibility issue and this code should be moved into the CombineFns class.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/coders/CoderRegistry.java [hunk 5]
```diff
@@ -235,7 +235,9 @@ public void registerCoderForType(TypeDescriptor<?> type, Coder<?> coder) {
...
```
Comment (lukecwik) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/coders/CoderRegistry.java]: `getCoderFromDescriptor` doesn't exist, do you mean the private `getCoderFromTypeDescriptor`?.

Also, this method has no replacement and is really meant to be copied into `ParDo` having `ParDo` own all that logic.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/coders/CoderRegistry.java [hunk 6]
```diff
@@ -277,7 +281,9 @@ public void registerCoderForType(TypeDescriptor<?> type, Coder<?> coder) {
...
```
Comment (lukecwik) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/coders/CoderRegistry.java]: There is no intended alternative, this is a visibility issue and this code should be copied into the CombineFnBase class.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/coders/CoderRegistry.java (hunk 5 above)
Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/coders/CoderRegistry.java]: Yea, I mean `getCoderFromTypeDescriptor`. Basically we do need at least one (and probably exactly one) public method where you can pass in a context of type-to-coder bindings. And the rest should be directed to use that.

Comment (lukecwik) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/coders/CoderRegistry.java]: Ken, have you seen enough demand for such a function in user based transforms that warrants visibility vs having a few independent copies?

Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/coders/CoderRegistry.java]: I don't think it can be implemented outside of the registry.

Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/coders/CoderRegistry.java]: Maybe if all the uses are just special cases that don't require real coder inference, that could be fine. But this method itself is quite simple so probably specialized versions would add complexity. I'm open to whatever, though.

Comment (lukecwik) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/coders/CoderRegistry.java]: If you don't believe that is possible, we should change the `@deprecated` comment to be something like:
`@deprecated This method is to change in an unknown backwards incompatible way once support for this functionality is refined.`

Similarly for the other CoderRegistry `@Internal/@Deprecated` methods.

File: runners/core-java/src/main/java/org/apache/beam/runners/core/InMemoryTimerInternals.java [hunk 7]
```diff
@@ -107,6 +107,9 @@ public void setTimer(StateNamespace namespace, String timerId, Instant target,
     setTimer(TimerData.of(timerId, namespace, target, timeDomain));
   }
 
+  /**
+   * @deprecated use {@link #setTimer(StateNamespace, String, Instant, TimeDomain)}.
```
Comment (alex-filatov) [inline runners/core-java/src/main/java/org/apache/beam/runners/core/InMemoryTimerInternals.java]: Unfortunately doesn't work. Only main description, `@param`, `@returns`, `@throws` are inheritable.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/coders/CoderRegistry.java (hunk 4 above)
Comment (alex-filatov) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/coders/CoderRegistry.java]: So purpose of the deprecation here is to attract attention to the `@Internal` annotation? In this case I propose to remove deprecation for consistency with the remaining codebase.

File: runners/core-java/src/main/java/org/apache/beam/runners/core/InMemoryTimerInternals.java (hunk 7 above)
Comment (kennknowles) [inline runners/core-java/src/main/java/org/apache/beam/runners/core/InMemoryTimerInternals.java]: Ah, too bad. Thanks for clarifying.

File: runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/PTransformTranslation.java [hunk 8]
```diff
@@ -55,7 +55,9 @@
   // Less well-known. And where shall these live?
   public static final String WRITE_FILES_TRANSFORM_URN = "urn:beam:transform:write_files:0.1";
 
+  // CHECKSTYLE.OFF: MissingDeprecated
   @Deprecated
+  // CHECKSTYLE.ON: MissingDeprecated
```
Comment (kennknowles) [inline runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/PTransformTranslation.java]: This is another case of the `CreatePCollectionView` transform being deprecated entirely.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/coders/CoderRegistry.java (hunk 5 above)
Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/coders/CoderRegistry.java]: Sounds good to me

File: sdks/java/core/src/main/java/org/apache/beam/sdk/coders/CoderRegistry.java (hunk 4 above)
Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/coders/CoderRegistry.java]: See above comment - just deprecated to indicate that it is going to change. So even though it is `@Internal` (and the javadoc should be loud about this, like in other places) it is helpful for contributors to see their IDEs strikethrough the method.

**Comment by alex-filatov** [comment]
All done. Please take a look.

**Comment by jkff** [comment]
This PR broke Dataflow runner because it deleted this code:
https://github.com/apache/beam/pull/3531/files#diff-59c03929b0864dc394262c8c6d19880bL240
https://github.com/apache/beam/pull/3531/files#diff-2d035c2598b124fa14e146e7af35905dL42
I'm going to reintroduce both of these. The comments on that code should have been more clear, like "Do not delete this without running Dataflow ValidatesRunner tests" :)

**Comment by alex-filatov** [comment]
Sorry for this. I couldn't find usage of this code in the repo, so assumed it was safe to delete.

**Comment by kennknowles** [comment]
No need to apologize. Your PR was just fine. It is on us to either make it a core part of the API or document why it exists despite being unused. Since it was leftover for compatibility, one way to make it obvious would be with a test in `DataflowRunner`.

And [partial] rollbacks are painless - don't view it as indicating any fault. A large number of our tests do not run until postcommit (to avoid massive amounts of computation) so it is expected for such things to happen fairly often.

**Pull request closed without merging**